Sources/RemoteMessaging/RemoteMessagingStore.swift

```diff
@@ -399,59 +400,67 @@ extension RemoteMessagingStore {
 
 extension RemoteMessagingStore {
 
-    private func save(remoteMessage: RemoteMessageModel, in context: NSManagedObjectContext) {
-        context.performAndWait {
-            let remoteMessageManagedObject = RemoteMessageManagedObject(context: context)
+    private func addOrUpdate(remoteMessage: RemoteMessageModel, in context: NSManagedObjectContext) {
+        if let scheduledRemoteMessage = RemoteMessageUtils.fetchScheduledRemoteMessage(in: context), scheduledRemoteMessage.id == remoteMessage.id {
+            scheduledRemoteMessage.message = RemoteMessageMapper.toString(remoteMessage) ?? ""
+        } else {
+            markScheduledMessagesAsDone(in: context)
+
+            let remoteMessageManagedObject = RemoteMessageUtils.fetchRemoteMessage(with: remoteMessage.id, in: context)
+            ?? RemoteMessageManagedObject(context: context)
 
-            remoteMessageManagedObject.id = remoteMessage.id
             remoteMessageManagedObject.message = RemoteMessageMapper.toString(remoteMessage) ?? ""
             remoteMessageManagedObject.status = NSNumber(value: RemoteMessageStatus.scheduled.rawValue)
-            remoteMessageManagedObject.shown = false
 
-            do {
-                try context.save()
-            } catch {
-                errorEvents?.fire(.saveMessageFailed, error: error)
-                os_log("Failed to save remote message entity: %@", log: log, type: .error, error.localizedDescription)
+            if remoteMessageManagedObject.isInserted {
+                remoteMessageManagedObject.id = remoteMessage.id
+                remoteMessageManagedObject.shown = false
```

Comment:
We only set `shown` to `false` when it's a new message. For reused message IDs we don't clear `shown` flag (this is in line with [Android implementation](https://github.com/duckduckgo/Android/pull/4760/files#diff-888a74c291c2d17fba08c8c6901c3493107ed4b362a313dc8acc324d51e5d71aR79-R80)).